### ISSUE TITLE
OpenSSL issue and possible use of urandom

### ISSUE THREAD

**Issue by daniels1006** [James1345/django-rest-knox#41]
Hi @James1345 and other contributors, first thanks for the great app!

I use Knox in production on several projects. This time I need to deploy one of those in a Linux shared server in a hosting provider (WebFaction). The problem is when I login I always receive the same token no matter the user credentials I send. Naturally, when trying to login a second time I get:

```
"(1062, "Duplicate entry '1080f871e39cc839e5bcc9f852f9a8f3ddc03cf7e72e9fd1d6e4a71d7e74936f' for key 'PRIMARY'")".
```

I noticed the problem is with the random bytes generation. When Knox tries to generate salt and token it only gets a bunch of zeros from `OpenSSL.rand.bytes`. So when it uses the zeros to digest the token, the result is always the same. This generates a database error since token field is, of course, unique.

As I needed to deploy the project fast I started to mess around with Knox code and found that if I replace the `OpenSSL.rand.bytes` with `os.urandom` it solves the problem. Is also good to mention both `os.urandom` and `OpenSSL.rand.bytes` work fine in the Python terminal of the same server.

In the documentation of Knox is said that "DRF Tokens are generated with os.urandom, which is not cryptographically secure. Knox uses OpenSSL to provide tokens." But the documentation of [pyOpenSSL rand](http://www.pyopenssl.org/en/stable/api/rand.html) module says "Functions from this module shouldn’t be used. Use urandom instead." pointing the [this](https://sockpuppet.org/blog/2014/02/25/safely-generate-random-numbers/) article.

So could some one clarify if is it really worse to use `os.random` and why?

**Comment by James1345** [comment]
Sorry it's taken me a while to get back to this. 

There are two reasons for this, the first is that when Knox was first being put together that warning was not here in such large letters. As always technology has moved on fairly quickly.

The other is one of Platform specifics. While /dev/urandom is preferable to OpenSSL random numbers on *nix devices, there is nothing to stop someone deploying Knox on a Windows based device, where `os.random` uses the system PRNG (CryptGenRand). The specifications for this are not available, it may not be a secure generator. At the time, it was considered that using Openssl seeded with system random would be secure on all systems, without any of that messy system detection code.

It is the consideration that most often is reconsidered. I think that it might be worth putting in a setting to let this be changed per-deployment, as both options have strengths.

Hope this clarifies the decisions made.

As for your specific issue, I cannot be sure why you are always getting the same from rand.bytes. It seems to be tied to the system you are deploying to. I would be interested to see if urandom is in any way deterministic on that server, as it is meant to be used to seed OpenSSL on creation of the generator.

**Comment by daniels1006** [comment]
Sorry I took long to get back too. 

It's clear now the rationale of the decision and it makes sense. Technology moves so fast that decisions that once were natural suddenly becomes not obvious anymore...

The system I'm deploying to is a fairly new machine running CentOS 7 (Linux Kernel 3.10). As I understood from what I've read [here](http://www.2uo.de/myths-about-urandom/) and in the article mentioned before, in this kind of environment, it would be potentially unsafe to use `/dev/urandom` in two cases: 

1. During the first boot of the machine with freshly installed OS;

2. If it's a VM that was cloned or rewinded to a previous state.

None of those is my case so I have no reason to think `/dev/urandom` is deterministic on that server.

Back to Knox design, introducing a setting seems like a good idea but it may not be necessary. I did some research about `CryptGenRandom` and it seems to be the prefered way of getting cryptographically safe random bytes in Windows (as `/dev/urandom` is in *nix), as It's said [here](https://msdn.microsoft.com/pt-br/library/windows/desktop/aa379942(v=vs.85).aspx).

I'm by no means an expert neither did an extensive and carefull research on this issue but I'm pretty convinced that just using `os.random` instead of `OpenSSL.rand.bytes` would be safer and eliminate a dependency on the project.

If you find that makes sense I can submit a pull request to do that (or at least the setting thing). That'd be usefull to me as I'd be able to get my project back using the oficial version of Knox.
